hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java

```diff
@@ -1404,15 +1403,38 @@ private JobTaskItemByHostMap generateTaskCountByHost(List<HostState> hosts, Iter
     }
 
     private List<JobTaskMoveAssignment> pushTasksOntoHost(HostState host, Iterable<HostState> heavyHosts) {
+        int moveLimit = config.getTasksMovedFullRebalance();
         MoveAssignmentList moveAssignments = new MoveAssignmentList(spawn, taskSizer);
+        List<String> targetHosts = new ArrayList<>();
+
+        List<HostState> lightHostList = new ArrayList<>();
+        lightHostList.add(host);
+        double byteLimitFactor = 1;
+
         for (HostState heavyHost : heavyHosts) {
-            double byteLimitFactor =
-                    1 - ((double) moveAssignments.getBytesUsed() / config.getBytesMovedFullRebalance());
-            moveAssignments.addAll(pushTasksOffHost(heavyHost, Collections.singletonList(host), true, byteLimitFactor,
-                                                    config.getTasksMovedFullRebalance(), true));
+            if (moveAssignments.size() >= moveLimit || !canReceiveNewTasks(host) || moveAssignments.getBytesUsed() >= config.getBytesMovedFullRebalance()) {
+                break;
+            }
+            List<JobTaskMoveAssignment> assignments = pushTasksOffHost(heavyHost, lightHostList,true, byteLimitFactor, 1, true);
+
+            // Don't add tasks from the same job which were on different hosts
+            for(JobTaskMoveAssignment assignment : assignments) {
+                if(!moveAssignments.containsJob(assignment.getJobKey().getJobUuid())) {
+                    moveAssignments.add(assignment);
+                    // Update byteLimitFactor with available byte limit
+                    byteLimitFactor = 1 - (moveAssignments.getBytesUsed() / config.getBytesMovedFullRebalance());
+                    targetHosts.add(assignment.getTargetUUID());
+                } else if(!targetHosts.contains(assignment.getTargetUUID())) {
+                    // Target host wrongly marked as recently replicated to
+                    // as we do not accept this moveAssignment
+                    recentlyReplicatedToHosts.invalidate(assignment.getTargetUUID());
```

Comment:
Now that I'm thinking about it, `assignment.getTargetUUID` is always `host` because we are moving tasks to it. Invalidation here is only necessary if we end up not moving any task onto this host. Even in that case, because `recentlyReplicatedToHosts` is a cache with a 15-minute expiration, the host may be in it from a previous rebalance call, so just invalidating it may be incorrect. In other words, you should invalidate only if it is not in `recentlyReplicatedToHosts` previously. At this point, I feel the code is becoming more complex than it needs to be, plus after a quick look, `recentlyReplicatedToHosts` seems to be only marked in `pushTasksOffHost`, which begs the question - why aren't we marking hosts in other places where we move tasks?